### PR TITLE
First Version Sprite Renderer(Rect mode)

### PR DESCRIPTION
**Pull request by LinYunMo**
Re: https://github.com/cocos/3d-tasks/issues/9127
Depend: https://github.com/cocos/engine-extensions/pull/70
Re: https://github.com/cocos/cocos-editor/pull/150

### CODE REVIEW

File: cocos/2d/framework/sprite-renderer.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    protected _size: Vec2 = new Vec2(); // todo for sliced & tiled
+
+    private _model: SpriteModel | null = null;
+
+    public __preload () {
+    }
+
+    public onLoad () {
+        if (this._spriteFrame) {
+            this._spriteFrame.mesh.initialize();
```
Comment (SantyWang) [inline cocos/2d/framework/sprite-renderer.ts:92]: 这里不太好，理论上 spriteFrame 的数据是在 spriteFrame 内部处理好的，不应该让组件去初始化

File: cocos/2d/framework/sprite-renderer.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    set spriteFrame (value) {
+        if (this._spriteFrame === value) {
+            return;
+        }
+
+        const lastSprite = this._spriteFrame;
+        this._spriteFrame = value;
+        if (this._spriteFrame) {
+            const mesh = this._spriteFrame.mesh;
+            mesh.initialize();
```
Comment (SantyWang) [inline cocos/2d/framework/sprite-renderer.ts:61]: 不应该在组件内去初始化资源

File: editor/assets/effects/builtin-sprite-renderer.effect [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  in vec3 a_position;
+  in vec2 a_texCoord;
+
+  out vec2 uv0;
+
+  vec4 vert () {
+    vec4 pos = vec4(a_position, 1);
+
+    pos = cc_matProj * (cc_matView * cc_matWorld) * pos;
```
Comment (SantyWang) [inline editor/assets/effects/builtin-sprite-renderer.effect]: mvp 有矩阵的吧

File: editor/assets/effects/builtin-sprite-renderer.effect [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        depthTest: true
+        depthWrite: false
+      blendState:
+        targets:
+        - blend: true
+          blendSrc: src_alpha
+          blendDst: one_minus_src_alpha
+          blendDstAlpha: one_minus_src_alpha
+      rasterizerState:
+        cullMode: none
```
Comment (SantyWang) [inline editor/assets/effects/builtin-sprite-renderer.effect:18]: 这个其实也可以考虑下

File: cocos/2d/framework/sprite-renderer.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    @visible(false)
+    @serializable
+    protected _size: Vec2 = new Vec2(); // todo for sliced & tiled
+
+    private _model: SpriteModel | null = null;
+
+    public __preload () {
+    }
+
+    public onLoad () {
```
Comment (SantyWang) [inline cocos/2d/framework/sprite-renderer.ts:90]: onEnable seems better

File: cocos/2d/framework/sprite-renderer.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+/**
+ * @packageDocumentation
+ * @module 2D
```
Comment (pandamicro) [inline cocos/2d/framework/sprite-renderer.ts]: No need to add package documentation any more

File: cocos/2d/assets/sprite-frame.ts [hunk 7]
```diff
@@ -41,25 +41,44 @@ import { Texture2D } from '../../core/assets/texture-2d';
...
 interface IVertices {
-    x: any;
-    y: any;
-    triangles: any;
-    nu: number[];
-    u: number[];
-    nv: number[];
-    v: number[];
+    rawPosition: Vec3[]; // 顶点的原始位置，像素值
+    positions: number[]; // 顶点的位置，受几个值影响的模型空间值 // 计算
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: Always use English comment, also comment the internal layout of the data, like 

x0, y0, x1, y1 ...

File: cocos/2d/assets/sprite-frame.ts [hunk 8]
```diff
@@ -41,25 +41,44 @@ import { Texture2D } from '../../core/assets/texture-2d';
...
 
 interface IVertices {
-    x: any;
-    y: any;
-    triangles: any;
-    nu: number[];
-    u: number[];
-    nv: number[];
-    v: number[];
+    rawPosition: Vec3[]; // 顶点的原始位置，像素值
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: Why we need rawPositions

File: cocos/2d/assets/sprite-frame.ts [hunk 9]
```diff
@@ -1243,26 +1316,41 @@ export class SpriteFrame extends Asset {
...
+                    uv: vertices.uv,
+                    nuv: vertices.nuv,
+                    minPos: new Vec3(vertices.minPos.x, vertices.minPos.y, vertices.minPos.z),
+                    maxPos: new Vec3(vertices.maxPos.x, vertices.maxPos.y, vertices.maxPos.z),
+                };
+            }
+            this.vertices.rawPosition.length = 0;
+            const rawPosition = vertices.rawPosition;
+            for (let i = 0; i < rawPosition.length; i += 3) {
+                temp_vec3.set(rawPosition[i], rawPosition[i + 1], rawPosition[i + 2]);
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: `new Vec3` to replace `set` plus `clone`

Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: `new Vec3` to replace `set` plus `clone`

File: cocos/2d/assets/sprite-frame.ts [hunk 10]
```diff
@@ -1243,26 +1316,41 @@ export class SpriteFrame extends Asset {
...
+        // const v = this.vertices;
+        // sp.vertices = v ? {
+        //     x: v.x,
+        //     y: v.y,
+        //     triangles: v.triangles, /* need clone ? */
+        //     nu: v.nu?.slice(0),
+        //     u: v.u?.slice(0),
+        //     nv: v.nv?.slice(0),
+        //     v: v.v?.slice(0),
+        // } : null as any;
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: vertices need to be cloned

File: cocos/2d/assets/sprite-frame.ts [hunk 11]
```diff
@@ -41,25 +41,44 @@ import { Texture2D } from '../../core/assets/texture-2d';
...
-    x: any;
-    y: any;
-    triangles: any;
-    nu: number[];
-    u: number[];
-    nv: number[];
-    v: number[];
+    rawPosition: Vec3[]; // 顶点的原始位置，像素值
+    positions: number[]; // 顶点的位置，受几个值影响的模型空间值 // 计算
+    triangles: number[]; // IB
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts]: `indexes` seems more appropriate

File: cocos/2d/assets/sprite-frame.ts [hunk 12]
```diff
@@ -1316,6 +1410,140 @@ export class SpriteFrame extends Asset {
...
+        const temp_vec3 = new Vec3(units, units, 1);
+        temp_matrix.scale(temp_vec3);
+        const PosX = -(this._pivot.x - 0.5) * this.rect.width * units;
+        const PosY = -(this._pivot.y - 0.5) * this.rect.height * units;
+        temp_vec3.set(PosX, PosY, 0);
+        temp_matrix.translate(temp_vec3);
+        const vertices = this.vertices!;
+
+        for (let i = 0; i < vertices.rawPosition.length; i++) {
+            const pos = vertices.rawPosition[i];
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts:1497]: Can we just share `rawPositions` and `positions`, convert each position and store back
But we might need to save rawPositions in editor, it's not a big deal, just want to save memory foot print in runtime

File: cocos/2d/framework/sprite-model.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import { Sampler } from '../../core/gfx/base/states/sampler';
+import { Texture } from '../../core/gfx/base/texture';
+import { ModelLocalBindings } from '../../core/pipeline/define';
+import { Model } from '../../core/renderer/scene/model';
+
+export class SpriteModel extends Model {
```
Comment (pandamicro) [inline cocos/2d/framework/sprite-model.ts]: Why SpriteModel is necessary, just update manually using Model API is enough

File: cocos/2d/framework/sprite-renderer.ts [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+    @serializable
+    protected _flipY = false;
+    @visible(false)
+    @serializable
+    protected _size: Vec2 = new Vec2(); // todo for sliced & tiled
+
+    private _model: SpriteModel | null = null;
+
+    public __preload () {
+    }
```
Comment (pandamicro) [inline cocos/2d/framework/sprite-renderer.ts]: No need to write empty extended function

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos/2d/assets/sprite-frame.ts (hunk 8 above)
Comment (LinYunMo) [reply cocos/2d/assets/sprite-frame.ts]: when custom change some attribute affected vertex position，wo need rawPositon to calculation

File: cocos/2d/assets/sprite-frame.ts [hunk 15]
```diff
@@ -36,25 +36,44 @@ import { Texture2D } from '../../core/assets/texture-2d';
...
 const INSET_LEFT = 0;
 const INSET_TOP = 1;
 const INSET_RIGHT = 2;
 const INSET_BOTTOM = 3;
+const temp_vec3 = new Vec3();
+const temp_matrix = new Mat4();
+
+enum MeshType {
+    RECT = 0,
+    POLYGON = 1, // todo
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts:52]: TODO what? please mark detail for all todo

**Review by SantyWang — APPROVED** [review]

File: cocos/2d/assets/sprite-frame.ts [hunk 16]
```diff
@@ -487,6 +510,21 @@ export class SpriteFrame extends Asset {
...
 
+    get pixelsToUnit () {
+        return this._pixelsToUnit;
+    }
+
+    get pivot () {
+        return this._pivot;
+    }
+
+    get mesh () {
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts:521]: API doc for all new public APIs

File: cocos/2d/assets/sprite-frame.ts [hunk 17]
```diff
@@ -537,6 +575,28 @@ export class SpriteFrame extends Asset {
 
     protected _packable = true;
 
+    protected _pixelsToUnit = 100;
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts:578]: Why set default value to 100

File: cocos/2d/assets/sprite-frame.ts [hunk 18]
```diff
@@ -1298,6 +1381,9 @@ export class SpriteFrame extends Asset {
         }
 
         this._checkPackable();
+        if (this._mesh) {
+            this.updateMesh();
```
Comment (pandamicro) [inline cocos/2d/assets/sprite-frame.ts:1385]: Do we need to update mesh here ? Update mesh have high memory footprint

File: cocos/2d/framework/sprite-renderer.ts [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    SLICED = 1,
+    TILED = 2,
+}
+
+@ccclass('cc.SpriteRenderer')
+@help('i18n:cc.SpriteRenderer')
+@executionOrder(100)
+@menu('2D/SpriteRenderer')
+@executeInEditMode
+export class SpriteRenderer extends ModelRenderer {
```
Comment (pandamicro) [inline cocos/2d/framework/sprite-renderer.ts:46]: API doc

**Review by pandamicro — APPROVED** [review]

File: cocos/2d/assets/sprite-frame.ts (hunk 17 above)
Comment (LinYunMo) [reply cocos/2d/assets/sprite-frame.ts:578]: Experience Value，users need to adjust the size of the resources themselves

File: cocos/2d/assets/sprite-frame.ts (hunk 18 above)
Comment (LinYunMo) [reply cocos/2d/assets/sprite-frame.ts:1385]: Yes it is required, but it does not have an impact if the SpriteRenderer is not used

**Pull request merged**